server2/app.py

```diff
@@ -0,0 +1,333 @@
+            initial_path=request.full_path
+        )
+        cookie.add_jwt(oauth_state, state_token_params)
+        return cookie.populate_resp(make_response(redirect(authorization_url)))
+    return wrapped
+
+
+@app.route('/', defaults={'key': None})
+@app.route("/<path:key>")
+@with_auth
+def index(user, key=None):
+    logger.debug('In index')
+    s3 = boto3.client('s3')
+
+    try:
+        logger.debug("s3 key argument: %s", key)
+        if not key:
+            key = default_object
+        logger.debug("looking in bucket [%s] for key [%s]", BUCKET_NAME, key)
+        response = s3.get_object(
+            Bucket=BUCKET_NAME,
+            Key=key
+        )
+        logger.debug(
+            'Retrieved object with content type: %s',
+            response['ContentType']
+        )
+    except botocore.exceptions.ClientError as e:
+        if e.response['Error']['Code'] == "404":
+            logger.warning("The object does not exist.")
+        else:
+            logger.error(e)
+    return Response(
+        str(response['Body'].read(), 'utf-8'),
+        mimetype=response['ContentType']
+    )
+
+    return key
+
+
+@app.route('/callback')
+def callback():
+    logger.debug('In callback')
+    """ Step 3: Retrieving an access token.
+
+    The user has been redirected back from the provider to your registered
+    callback URL. With this redirection comes an authorization code
+    included in the redirect URL.
+    We will use that to obtain an access token.
+    """
+    jwt = get_jwt_token(request.cookies, oauth_state)
+    logger.info('got jwt token: %s', jwt)
+    logger.info('looking for: %s', oauth_state_key)
+    if oauth_state_key not in jwt:
+        # something is wrong with the state token,
+        # so redirect back to start over.
+        logger.warn('no state passed to openid callback')
+        return redirect(url_for('.index'))
+
+    initial_path = url_for('.index')
+    if oauth_initial_path_key in jwt:
+        initial_path = jwt[oauth_initial_path_key]
+
+    logger.debug('initial_path: %s', initial_path)
+    client = OAuth2Session(
+        client_id,
+        state=jwt.get('oauth_state'),
+        redirect_uri=redirect_uri
+    )
+
+    token = client.fetch_token(
+        token_endpoint,
+        client_secret=client_secret,
+        authorization_response=base_uri + request.full_path
+    )
+    userinfo = client.get(userinfo_endpoint).json()
+    # Save the token
+    cookie = Cookie()
+    exp_seconds = 60 * 60 * 4
+    if "ExpiresIn" in token:
+        exp_seconds = token['ExpiresIn']
+    jwt_token_params = jwt_params(
+        exp_seconds=exp_seconds,
+        oauth_token=token,
+        **userinfo
+    )
+    cookie.add_jwt(oauth_token, jwt_token_params)
+    cookie.remove_jwt(oauth_state)
+    print('initial_path', initial_path)
+    return cookie.populate_resp(make_response(redirect(initial_path)))
+
+
+if __name__ == "__main__":
+    app.secret_key = os.urandom(24)
```

Comment:
Does this break things for users if you redeploy say twice in a day, and their session information stays the same?

Reply:
This app doesn't use the session store, so I should just delete this. Good catch.